### PR TITLE
Session

### PR DESCRIPTION
**Pull request by CrisVega08**
## Summary

- Create Route for /users/sessions
- Controller Sign In
- Services for sign In
- Test for new route

image

### CODE REVIEW

File: app/controllers/user.js [hunk 1]
```diff
@@ -9,3 +10,13 @@ exports.signUp = (req, res, next) =>
       res.status(201).send(serializer(response));
     })
     .catch(next);
+
+exports.signIn = (req, res, next) =>
+  signIn(req.body)
+    .then(token =>
+      res
+        .header({ [config.common.session.headerName]: token })
+        .status(201)
```
Comment (jersonlopez) [inline app/controllers/user.js:19]: I think that this status should be 200, 201 status is for `created`. what do you think?

File: app/middlewares/errors.js [hunk 2]
```diff
@@ -7,7 +7,8 @@ const statusCodes = {
   [errors.DATABASE_ERROR]: 503,
   [errors.DEFAULT_ERROR]: 500,
   [errors.EXIST_REGISTER]: 409,
-  [errors.SCHEMA_VALIDATION]: 422
+  [errors.SCHEMA_VALIDATION]: 422,
+  [errors.WRONG_CREDENTIALS]: 422
```
Comment (jersonlopez) [inline app/middlewares/errors.js]: 422 status is for params validation, no for invalid credentials, check HTTP status codes please

File: app/services/user.js [hunk 3]
```diff
@@ -14,3 +15,17 @@ exports.signUp = async body => {
...
 };
+
+exports.signIn = async ({ email, password }) => {
+  try {
+    const user = await User.getUser({ email });
+    if (user && (await compareHash(password, user.dataValues.password))) {
+      logger.info(`${user.dataValues.email} sign in correctly`);
+      return generaToken(user.dataValues);
+    }
+    logger.info(`wrong sign in user: ${email}`);
```
Comment (jersonlopez) [inline app/services/user.js:26]: typo(?)

File: app/utils/session.js [hunk 4]
```diff
@@ -0,0 +1,14 @@
+const bcrypt = require('bcryptjs');
+const jwt = require('jsonwebtoken');
+const config = require('../../config');
+
+const salt = bcrypt.genSaltSync(parseInt(config.common.session.salt));
```
Comment (jersonlopez) [inline app/utils/session.js]: here you can set a default value if the salt is no set in the env variables:
```
config.common.session.salt || 10;
```

File: app/utils/session.js [hunk 5]
```diff
@@ -0,0 +1,14 @@
...
+
+exports.hashPassword = password => bcrypt.hashSync(password, salt);
+
+exports.compareHash = (password, hash) => bcrypt.compareSync(password, hash);
+
+exports.generaToken = ({ id }) => {
+  const { secret, expiresIn } = config.common.session;
```
Comment (jersonlopez) [inline app/utils/session.js]: you can group all config and do the requirement in the top of file: 
```
const { secret, expiresIn, salt } = config.common.session;
```

File: config/index.js [hunk 6]
```diff
@@ -37,14 +37,18 @@ const config = {
       port: process.env.PORT
     },
     session: {
-      header_name: 'authorization',
+      headerName: 'authorization',
```
Comment (jersonlopez) [inline config/index.js:40]: thanks for this 👍

File: test/factory/user.js [hunk 7]
```diff
@@ -1,7 +1,8 @@
 const { factory } = require('factory-girl');
-const { factoryWithCustomizedValue } = require('./factory_by_models');
 const { mockUser } = require('../constants');
+const { hashPassword } = require('../../app/utils/session');
+const { factoryByModel } = require('./factory_by_models');
 
-factoryWithCustomizedValue('User', 'email', mockUser.email);
+factoryByModel('User', { ...mockUser, password: () => hashPassword(mockUser.password) });
```
Comment (jersonlopez) [inline test/factory/user.js:6]: good solution 🥇

File: app/routes.js [hunk 8]
```diff
@@ -1,16 +1,17 @@
...
 // user controller
-const { signUp } = require('./controllers/user');
+const { signUp, signIn } = require('./controllers/user');
 
 exports.init = app => {
   app.get('/health', healthCheck);
   app.get('/albums', getAlbums);
   app.get('/albums/:id/photos', getPhotos);
   app.post('/users', paramsValidator(userSchema), signUp);
+  app.post('/users/sessions', paramsValidator(signInSchema), signIn);
```
Comment (jersonlopez) [inline app/routes.js:16]: I think this is a GET request because they're requesting information to us, what do you think?

**Review by jersonlopez — CHANGES_REQUESTED** [review]

File: app/controllers/user.js (hunk 1 above)
Comment (CrisVega08) [reply app/controllers/user.js:19]: ok, I use 201 status code because the service is going to create a token

File: app/services/user.js (hunk 3 above)
Comment (CrisVega08) [reply app/services/user.js:26]: I don't understand your comment can be more specific

File: app/routes.js (hunk 8 above)
Comment (CrisVega08) [reply app/routes.js:16]: if we use a get, i have to send params on query and I think is weird. But tell me how you do that?

**Review by jersonlopez — APPROVED** [review]

**Pull request merged**